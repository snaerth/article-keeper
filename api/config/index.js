// Eviromental variables
import dotenv from 'dotenv';

dotenv.config();

const config = {
  // Application name
  APPLICATION_NAME: process.env.APPLICATION_NAME || 'Application name',

  // Uploads root folder
  UPLOADS_ROOT: process.env.UPLOADS_ROOT || './uploads/',

  // Defaults
  NODE_ENV: process.env.NODE_ENV || 'development',

  // Api Protocol
  API_PROTOCOL: process.env.API_PROTOCOL || 'http',

  // Api Port
  API_PORT: process.env.API_PORT || 3030,

  // Api Host
  API_HOST: process.env.API_HOST || 'localhost',

  // Protocol
  PROTOCOL: process.env.PROTOCOL || 'http',

  // Port
  PORT: process.env.PORT || 3000,

  // Host
  HOST: process.env.HOST || 'localhost',

  // Database url
  DB_URL: process.env.DB_URL || 'mongodb://localhost/application',

  // Database test url
  TEST_DB_URL: 'mongodb://localhost/test',

  // Database username
  DB_USERNAME: process.env.DB_USERNAME || 'username',

  // Database password
  DB_PASSWORD: process.env.DB_PASSWORD || 'password',

  // Passport jwt secret
  JWT_SECRET: process.env.JWT_SECRET || 'secret',

  // Express session
  SESSION_SECRET: process.env.SESSION_SECRET || 'secret',

  // Email username
  EMAIL_USERNAME: process.env.EMAIL_USERNAME || '',

  // Email password
  EMAIL_PASSWORD: process.env.EMAIL_PASSWORD || '',

  // Facebook client id
  FACEBOOK_CLIENT_ID: process.env.FACEBOOK_CLIENT_ID || '',

  // Facebook client secret
  FACEBOOK_CLIENT_SECRET: process.env.FACEBOOK_CLIENT_SECRET || '',

  // Twitter consumer key
  TWITTER_CONSUMER_KEY: process.env.TWITTER_CONSUMER_KEY || '',

  // Twitter consumer secret
  TWITTER_CONSUMER_SECRET: process.env.TWITTER_CONSUMER_SECRET || '',

  // Google client id
  GOOGLE_CLIENT_ID: process.env.GOOGLE_CLIENT_ID || '',

  // Google client secret
  GOOGLE_CLIENT_SECRET: process.env.GOOGLE_CLIENT_SECRET || '',
};

export default config;
