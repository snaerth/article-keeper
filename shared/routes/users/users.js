import React from 'react';
import Helmet from 'react-helmet';
// Components
import Users from '../../components/users';
import Container from '../../components/common/container';

const UsersRoute = () => (
  <div>
    <Helmet title="Users" />
    <Container className="pt25">
      <Users />
    </Container>
  </div>
);

export default UsersRoute;
