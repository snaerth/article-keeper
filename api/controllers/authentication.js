import jwt from 'jwt-simple';
import crypto from 'crypto';
import formidable from 'formidable';
import path from 'path';
import uuid from 'uuid/v1';
import { validateEmail } from '../services/utils';
import sendMail from '../services/mailService';
import { resizeImage } from '../services/imageService';
import { checkFileAndDelete } from '../services/fileService';
import User from '../models/user';
import config from '../config';

// VARIABLES
const { PORT, HOST } = config;

/**
 * Checks if user exist by email.
 * If user is found notify that email is in use
 *
 * @param {String} email
 * @returns {Promise}
 * @author Snær Seljan Þóroddsson
 */
async function checkUserByEmail(email) {
  try {
    // See if user with given email exists
    const user = await User.findOne({
      email,
    });
    // If a user does exist, return error
    if (user) {
      throw new Error('Email is in use');
    }

    return Promise.resolve();
  } catch (error) {
    throw new Error(error);
  }
}

/**
 * Validates email, password name, etc. from post request
 * If error then return error string
 *
 * @param {String} email
 * @param {String} password
 * @param {String} newPassword
 * @param {String} name
 * @param {String} dateOfBirth
 * @param {Object} res
 * @returns {String} error
 * @author Snær Seljan Þóroddsson
 */
export function validateSignup({
  email,
  password,
  newPassword,
  name,
  dateOfBirth,
}) {
  // Check if email, password or message exist in request
  if (!email || !password || !name) {
    return 'You must provide name, email and password';
  }
  // Validate email
  if (!validateEmail(email)) {
    return `${email} is not a valid email`;
  }

  // Check if password length is longer then 6 characters
  if (password.length < 6) {
    return 'Password must be of minimum length 6 characters';
  }
  // Check if password contains one number and one uppercase letter
  if (!/[0-9]/.test(password) || !/[A-Z]/.test(password)) {
    return 'Password must contain at least one number (0-9) and one uppercase letter (A-Z)';
  }

  if (newPassword) {
    // Check if password length is longer then 6 characters
    if (newPassword.length < 6) {
      return 'New password must be of minimum length 6 characters';
    }
    // Check if password contains one number and one uppercase letter
    if (!/[0-9]/.test(newPassword) || !/[A-Z]/.test(newPassword)) {
      return 'New password must contain at least one number (0-9) and one uppercase letter (A-Z)';
    }
  }

  // Name has to have aleast two names
  if (!/^([^0-9]*)$/.test(name) || name.trim().split(' ').length < 2) {
    return 'Name has aleast two 2 names consisting of letters';
  }

  if (dateOfBirth && !Date.parse(dateOfBirth)) {
    return 'Date is not in valid format. Try DD.MM.YYYY';
  }

  return null;
}

/**
 * Removes properties from mongoose user schema object
 *
 * @param {Object} user - Mongoose user schema object
 * @param {Array} moreProps - Array of strings to remove from user object
 * @returns {Object} newUser
 */
export function removeUserProps(user, moreProps) {
  let delProps = ['__v', '_id', 'createdAt', 'password'];
  delProps = moreProps && moreProps.length
    ? delProps.concat(moreProps)
    : delProps;
  const newUser = user.toObject();

  for (let i = 0; i < delProps.length; i++) {
    const hasBarProperty = Object.prototype.hasOwnProperty.call(
      newUser,
      delProps[i],
    );

    if (hasBarProperty) {
      delete newUser[delProps[i]];
    }
  }

  return newUser;
}

/**
 * Gets token for user
 * @param {Object} user - User object
 * @returns {String} token - Newly created token
 */
export function tokenForUser(user) {
  const timestamp = new Date().getTime();
  return jwt.encode(
    {
      sub: user.id,
      iat: timestamp,
    },
    process.env.JWT_SECRET,
  );
}

/**
 * Create new user and save to database
 *
 * @param {Object} user - Mongoose user schema object
 * @param {Object} propsToDelArr - Array of properites to delete
 * @returns {Promise}
 * @author Snær Seljan Þóroddsson
 */
export async function saveUser(user) {
  if (user.constructor.name === 'model') {
    try {
      // Save user to database
      await user.save();
      return Promise.resolve(user);
    } catch (error) {
      throw new Error(error);
    }
  } else {
    throw new Error('Object is not a mongoose object');
  }
}

/**
 * Sign up route
 * Sign user up to system. If no errors create user
 *
 * @param {Object} req
 * @param {Object} res
 * @returns {Object} res
 * @returns {undefined}
 * @author Snær Seljan Þóroddsson
 */
export async function signup(req, res) {
  if (!req.body) {
    return res.status(422).send({ error: 'No post data found' });
  }

  const { email, password, name } = req.body;
  // Validate post request inputs
  const error = validateSignup({ email, password, name });

  if (!error) {
    try {
      // Check for if user exists by email
      await checkUserByEmail(email);
      const user = new User({ name, email, password });
      // Save user in database
      const data = await saveUser(user);
      // Remove unwanted props for client
      const newUser = removeUserProps(user);
      // Send response object with user token and user information
      return res.status(200).json({
        token: tokenForUser(data),
        ...newUser,
      });
    } catch (err) {
      return res.status(422).send({ error: err });
    }
  } else {
    return res.status(422).send({ error });
  }
}

/**
 * Find user by email. If user is found then return user
 *
 * @param {String} email
 * @returns {Promise}
 * @author Snær Seljan Þóroddsson
 */
async function findUserByEmail(email) {
  try {
    // See if user with given email exists
    const user = await User.findOne({
      email,
    });

    // If a user does exist, return error
    if (!user) {
      throw new Error('No user found');
    }

    return Promise.resolve(user);
  } catch (error) {
    throw new Error(error);
  }
}

/**
 * Updates user and save to database
 *
 * @param {Object} req
 * @param {Object} res
 * @returns {Object} res
 * @author Snær Seljan Þóroddsson
 */
export async function updateUser(req, res) {
  if (!req.body) return res.status(422).send({ error: 'No post data found' });
  if (!req.user) return res.status(422).send({ error: 'No user found' });

  const { email, password, newPassword, name, dateOfBirth, phone } = req.body;
  // Validate post request inputs
  const error = validateSignup({
    email,
    password,
    newPassword,
    name,
    dateOfBirth,
  });

  if (!error) {
    try {
      // Find user by email and populate user props
      const user = await findUserByEmail(email);
      user.email = email;
      user.password = password;
      user.name = name;
      user.dateOfBirth = dateOfBirth;
      user.phone = phone;

      return user.comparePassword(password, async (err, isMatch) => {
        if (err) {
          return Promise.reject({ error: err });
        }

        if (!isMatch) {
          return Promise.reject({
            error: 'Password does not match old password',
          });
        }

        user.password = newPassword;
        // Save new user to databases
        const updatedUser = await saveUser(user);
        // Remove unwanted props for client
        const newUser = removeUserProps(user);
        // Send response object with user token and user information
        return res.status(200).json({
          token: tokenForUser(updatedUser),
          ...newUser,
        });
      });
    } catch (err) {
      return res.status(422).send({ error: err });
    }
  } else {
    return res.status(422).send({ error });
  }
}

/**
 * Upload user image to file system
 *
 * @param {Object} req
 * @param {Object} res
 * @returns {Object} res
 * @author Snær Seljan Þóroddsson */
export function uploadUserImage(req, res) {
  const { email } = req.user;
  const form = formidable.IncomingForm({ uploadDir: './public/images/users/' });

  form.on('error', () =>
    res.status(500).send({ error: 'An error has occured with image upload' }),
  );

  form.parse(req, async (err, fields, files) => {
    const image = files.image;

    if (image) {
      const ext = path.extname(image.name);
      const fileName = uuid();
      const imgPath = form.uploadDir + fileName + ext;
      const thumbnailPath = `${form.uploadDir}${`${fileName}-thumbnail${ext}`}`;

      try {
        const user = await findUserByEmail(email);
        const { imageUrl, thumbnailUrl } = user;

        if (imageUrl) {
          await checkFileAndDelete(form.uploadDir + imageUrl);
        }

        if (thumbnailUrl) {
          await checkFileAndDelete(form.uploadDir + thumbnailUrl);
        }

        user.imageUrl = fileName + ext;
        await resizeImage(image.path, imgPath, 400);
        await resizeImage(image.path, thumbnailPath, 27);

        user.thumbnailUrl = `${fileName}-thumbnail${ext}`;
        await checkFileAndDelete(image.path);
        const updatedUser = await saveUser(user);
        // Remove unwanted props for client
        const newUser = removeUserProps(user);
        // Send response object with user token and user information
        return res.status(200).json({
          token: tokenForUser(updatedUser),
          ...newUser,
        });
      } catch (error) {
        return res.status(422).send({ error });
      }
    } else {
      return res.status(422).send({ error: 'Image required' });
    }
  });
}

/**
 * Signin route
 * If users is authenticated responde with a token
 *
 * @param {Object} req
 * @param {Object} res
 * @returns {undefined}
 * @author Snær Seljan Þóroddsson
 */
export function signin(req, res) {
  const user = req.user;
  let data = null;

  if (user) {
    const userObject = removeUserProps(user);

    data = {
      token: tokenForUser(user),
      ...userObject,
    };

    if (user.roles.includes('admin')) {
      data.role = 'admin';
    }
  }

  res.send(data);
}

/**
 * Generates uniq token
 *
 * @returns {Promise} promise - Token
 * @author Snær Seljan Þóroddsson
 */
async function createRandomToken() {
  try {
    // Create buffer
    const buffer = await crypto.randomBytes(20);
    const token = buffer.toString('hex');
    return Promise.resolve(token);
  } catch (error) {
    throw new Error(error);
  }
}

/**
 * Finds user by email,if user exist
 * set resetPasswordToken and resetPasswordExpires props
 * save those props to user
 *
 * @param {Object} props - Token for user and email
 * @returns {Promise} promise - user
 * @author Snær Seljan Þóroddsson
 */
async function attachTokenToUser({ token, email }) {
  const hourInSeconds = 60 * 60 * 1000; // 1 hour

  try {
    const user = User.findOne({
      email,
    });

    user.resetPasswordToken = token;
    user.resetPasswordExpires = Date.now() + hourInSeconds;

    // Save user to databases
    await saveUser(user);
    // Remove unwanted props for client
    return removeUserProps(user);
  } catch (error) {
    throw new Error(error);
  }
}

/**
 * Send reset password email
 *
 * @returns {Promise} promise - User
 * @author Snær Seljan Þóroddsson
 */
async function sendResetPasswordEmail({ url, email, name }) {
  const mailOptions = {
    to: email,
    subject: 'Password reset',
    text: 'Password reset',
    html: `
          <p>Hi ${name}</p>
          <p>We've received a request to reset your password. If you didn't make the request</p>
          <p>just ignore this email. Otherwise you can reset your password using this link:</p>
          <a href="http://${url}">Click here to reset your password</a>
          <p>Thank you.</p>
      `,
  };

  const { to, subject, text, html } = mailOptions;

  try {
    // Send email
    const info = await sendMail(to, subject, text, html);
    // Return info and email
    return Promise.resolve({ info, email });
  } catch (error) {
    throw new Error(error);
  }
}

/**
 * Forgot password route
 * TODO describe info
 *
 * @param {Object} req
 * @param {Object} res
 * @returns {undefined}
 * @author Snær Seljan Þóroddsson
 */
export async function forgotPassword(req, res) {
  const { email } = req.body;

  // Create token Save resetPasswordToken and resetPasswordExpires to user Send
  // email to user
  try {
    const token = await createRandomToken();
    const { user } = await attachTokenToUser({ token, email });
    const url = `${HOST}:${PORT}/reset/${token}`;
    const { name } = user;
    const data = await sendResetPasswordEmail({ url, email, name });
    return res.send(
      `An e-mail has been sent to ${data.email} with further instructions.`,
    );
  } catch (error) {
    return res
      .status(550)
      .send({ error: "Coundn't reset password at this time.", err: error });
  }
}

/**
 * Finds user, updates properties (password, resetPasswordToken, resetPasswordExpires)
 * and saves user with new propertys
 *
 * @param {String} token
 * @param {String} password
 * @returns {undefined}
 * @author Snær Seljan Þóroddsson
 */
async function updateUserPassword({ token, password }) {
  const hourInSeconds = 60 * 60 * 1000;

  try {
    const user = User.findOne({
      resetPasswordToken: token,
      resetPasswordExpires: {
        $gt: Date.now() - hourInSeconds,
      },
    });

    if (!user) {
      return Promise.reject({
        error: 'Password reset token is invalid or has expired.',
      });
    }

    user.password = password;
    user.resetPasswordToken = undefined;
    user.resetPasswordExpires = undefined;

    // Save user to databases
    await user.save();
    return Promise.resolve(user);
  } catch (error) {
    throw new Error(error);
  }
}

/**
 * Resets user password
 * Finds user by resetPasswordToken property and resetPasswordExpires
 *
 * @param {Object} req
 * @param {Object} res
 * @returns {undefined}
 * @author Snær Seljan Þóroddsson
 */
export async function resetPassword(req, res) {
  const token = req.params.token;
  const password = req.body.password;

  if (token && password) {
    try {
      const user = await updateUserPassword({ token, password });
      return res.send(
        `Success! Your password has been changed for ${user.email}.`,
      );
    } catch (error) {
      return res.send({ error: 'Password is invalid or token has expired.' });
    }
  } else {
    return res.send({ error: 'Token and password are required' });
  }
}

/**
 * Check whether user has admin roles
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Func} next
 * @returns {undefined
 * @author Snær Seljan Þóroddsson
 */
export function isAdmin(req, res, next) {
  if (req.user && req.user.roles.includes('admin')) {
    return next();
  }

  return res.status(401).send({ error: 'Unauthorized' });
}
