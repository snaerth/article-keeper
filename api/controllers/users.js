import mongoose from 'mongoose';
import jwt from 'jwt-simple';
import path from 'path';
import uuid from 'uuid/v1';
import User from '../models/user';
import { resizeImage } from '../services/imageService';
import createPaginationObject from '../services/paginationService';
import { checkFileAndDelete, renameFile } from '../services/fileService';
import dbSortValues from '../services/dbService';
import log from '../services/logService';
import parseDateYearMonthDay from '../utils/date';

import { removeUserProps, validateSignup } from './authentication';

// Sort values for sorting in db
const sortVals = dbSortValues(); // [-1,1,'asc','desc']
// Properties to select from user collection in db
const select = {
  name: 1,
  email: 1,
  createdAt: 1,
  updatedAt: 1,
  roles: 1,
  imageUrl: 1,
  thumbnailUrl: 1,
  dateOfBirth: 1,
  profile: 1,
  phone: 1,
  'facebook.email': 1,
  'facebook.name': 1,
  'twitter.email': 1,
  'twitter.name': 1,
  'google.email': 1,
  'google.name': 1,
};

/**
 * Gets token for user
 * @param {Object} user - User object
 * @returns {String} token - Newly created token
 */
export function tokenForUser(user) {
  const timestamp = new Date().getTime();
  return jwt.encode(
    {
      sub: user._id, // eslint-disable-line
      iat: timestamp,
    },
    process.env.JWT_SECRET,
  );
}

/**
 * Creates user roles array
 *
 * @param {Array} roles
 * @returns {Array}
 */
function createUserRoles(roles) {
  const arr = ['user'];
  if (Array.isArray(roles)) {
    if (roles.includes('admin')) {
      arr.push('admin');
    }

    if (roles.includes('superuser')) {
      arr.push('superuser');
    }
  }

  return arr;
}

/**
 * Check whether user has admin role
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Func} next
 * @author Snær Seljan Þóroddsson
 */
export function isAdmin(req, res, next) {
  if (req.user && req.user.roles.includes('admin')) {
    return next();
  }

  return res.status(401).send({ error: 'Unauthorized' });
}

/**
 * Check whether user has superuser role
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Func} next
 * @author Snær Seljan Þóroddsson
 */
export function isSuperUser(req, res, next) {
  const { user, user: { roles } } = req;

  if (user && (roles.includes('superuser') || roles.includes('admin'))) {
    return next();
  }

  return res.status(401).send({ error: 'Unauthorized' });
}

/**
 * Enriches pagination object with sort and other pagination properties
 *
 * @param {Object} pagination - { limit: Number, offset: Number }
 * @param {String|Number} name
 * @param {String|Number} email
 * @param {String|Number} dateOfBirth
 * @returns {Object} sort
 * @author Snær Seljan Þóroddsson
 */
function addToPaginationObject(pagination, name, email, dateOfBirth, createdAt) {
  const sort = {};
  if (sortVals.includes(name)) sort.name = name;
  if (sortVals.includes(email)) sort.email = email;
  if (sortVals.includes(dateOfBirth)) sort.dateOfBirth = dateOfBirth;
  if (sortVals.includes(createdAt)) sort.createdAt = createdAt;
  if (Object.keys(sort).length !== 0) {
    pagination.sort = sort;
  }

  return pagination;
}

/**
 * Find user by email. If user is found then return user
 *
 * @param {String} email
 * @returns {Promise}
 * @author Snær Seljan Þóroddsson
 */
export async function findUserByEmail(email) {
  return new Promise(async (resolve, reject) => {
    try {
      const $or = [
        { email },
        { 'facebook.email': email },
        { 'twitter.email': email },
        { 'google.email': email },
      ];

      // See if user with given email exists
      const user = await User.findOne({
        $or,
      });

      // If a user does exist, return error
      if (!user) {
        return reject('No user found');
      }

      return resolve(user);
    } catch (err) {
      log.error({ err }, 'Error finding user by email');
      return reject(err);
    }
  });
}

/**
 * Create new user and save to database
 *
 * @param {Object} user - Mongoose user schema object
 * @param {Object} propsToDelArr - Array of properites to delete
 * @returns {Promise}
 * @author Snær Seljan Þóroddsson
 */
export function saveUser(user) {
  return new Promise(async (resolve, reject) => {
    if (user.constructor.name === 'model') {
      try {
        // Save user to database
        const newUser = await user.save();
        log.info({ info: newUser }, `User ${newUser.name} saved in db`);
        return resolve(newUser);
      } catch (err) {
        log.error({ err }, 'Error saving user');
        return reject(err);
      }
    } else {
      log.error(
        { err: new Error('Object is not a mongoose object') },
        'Object is not a mongoose object',
      );

      return reject('Object is not a mongoose object');
    }
  });
}

/**
 * Upload user image to file system
 *
 * @param {Object} req
 * @param {Object} res
 * @returns {Object} res
 * @author Snær Seljan Þóroddsson
 */
export async function uploadUserImage(req, res) {
  const { email } = req.fields;
  if (!email) return res.status(422).send('No email found in post data');

  const image = req.files.image;

  if (image) {
    try {
      const user = await findUserByEmail(email);
      const uploadDir = './media/users/';
      const ext = path.extname(image.name);
      const fileName = uuid();
      // File system path to image
      const imgPath = uploadDir + fileName + ext;
      // File system path to thumbnail image
      const thumbnailPath = `${uploadDir}${`${fileName}-thumbnail${ext}`}`;
      const { imageUrl, thumbnailUrl } = user;

      await resizeImage(image.path, thumbnailPath, 27);
      await renameFile(image.path, imgPath);

      if (imageUrl) {
        await checkFileAndDelete(imageUrl);
      }

      if (thumbnailUrl) {
        await checkFileAndDelete(thumbnailUrl);
      }

      const $set = {
        imageUrl: uploadDir + fileName + ext,
        thumbnailUrl: `${uploadDir + fileName}-thumbnail${ext}`,
        updatedAt: new Date(),
      };

      // Find user by email and populate user props
      const updatedUser = await User.findOneAndUpdate({ email }, { $set }, { new: true }).select(
        select,
      );

      // Send response object with and user information
      return res.status(200).json(updatedUser);
    } catch (err) {
      log.error({ req, res, err }, 'Error uploading user image');
      return res.status(422).send({ error: err });
    }
  } else {
    return res.status(422).send({ error: 'Image required' });
  }
}

/**
 * Checks if user exist by email.
 * If user is found notify that email is in use
 *
 * @param {String} email
 * @returns {Promise}
 * @author Snær Seljan Þóroddsson
 */
export function checkUserByEmail(email) {
  return new Promise(async (resolve, reject) => {
    try {
      // See if user with given email exists
      const user = await User.findOne({
        email,
      });
      // If a user does exist, return error
      if (user) {
        return reject('Email is in use');
      }

      return resolve();
    } catch (err) {
      log.error({ err }, 'Error checking if user exist by email');

      return reject(err);
    }
  });
}

/**
 * Finds user by email,if user exist
 * set resetPasswordToken and resetPasswordExpires props
 * save those props to user
 *
 * @param {Object} props - Token for user and email
 * @returns {Promise} promise - user
 * @author Snær Seljan Þóroddsson
 */
export async function attachTokenToUser({ token, email }) {
  return new Promise(async (resolve, reject) => {
    const hourInSeconds = 60 * 60 * 1000; // 1 hour

    try {
      const user = await findUserByEmail(email);
      user.resetPasswordToken = token;
      user.resetPasswordExpires = Date.now() + hourInSeconds;

      // Save user to databases
      await saveUser(user);
      // Remove unwanted props for client
      return resolve(removeUserProps(user));
    } catch (err) {
      log.error({ err }, 'Error attaching token to user object');

      return reject(err);
    }
  });
}

/**
 * Finds user, updates properties (password, resetPasswordToken, resetPasswordExpires)
 * and saves user with new propertys
 *
 * @param {String} token
 * @param {String} password
 * @returns {Promise}
 * @author Snær Seljan Þóroddsson
 */
export async function updateUserPassword({ token, password }) {
  return new Promise(async (resolve, reject) => {
    const hourInSeconds = 60 * 60 * 1000;

    try {
      const user = User.findOne({
        resetPasswordToken: token,
        resetPasswordExpires: {
          $gt: Date.now() - hourInSeconds,
        },
      });

      if (!user) {
        return reject({
          error: 'Password reset token is invalid or has expired.',
        });
      }

      user.password = password;
      user.resetPasswordToken = undefined;
      user.resetPasswordExpires = undefined;

      // Save user to databases
      await user.save();
      // Log in db
      log.info({ info: user }, `Update password for user ${user.name}`);

      return resolve(user);
    } catch (err) {
      log.error({ err }, 'Error updating user password');

      return reject(err);
    }
  });
}

/**
 * Create user
 *
 * @param {Object} req
 * @param {Object} res
 * @returns {Object} res
 * @author Snær Seljan Þóroddsson
 */
export async function createUser(req, res) {
  if (!req.body) return res.status(422).send('No post data found');
  const { email, password, name, phone, roles, dateOfBirth } = req.body;

  // Validate post request inputs
  const validateError = validateSignup({
    email,
    password,
    newPassword: null,
    name,
    dateOfBirth,
    phone,
  });

  if (!validateError) {
    try {
      // Check for if user exists by email
      await checkUserByEmail(email);

      const user = new User({
        name,
        email,
        password,
        dateOfBirth,
        phone,
        roles: createUserRoles(roles),
      });
      // Save user in database
      await saveUser(user);
      // Remove unwanted props for client
      const newUser = removeUserProps(user);
      // Send response object with user token and user information
      return res.status(200).json({
        ...newUser,
      });
    } catch (err) {
      log.error({ req, res, err }, 'Error creating user');
      return res.status(422).send({ error: err });
    }
  } else {
    return res.status(422).send({ error: validateError });
  }
}

/**
 * Delete user
 *
 * @param {Object} req
 * @param {Object} res
 * @returns {Object} res
 * @author Snær Seljan Þóroddsson
 */
export async function deleteUser(req, res) {
  const { id } = req.params;
  if (!id) {
    return res.status(400).send({ error: 'No id found in query string' });
  }

  try {
    const user = await User.findByIdAndRemove(id);
    if (user) {
      // Log in db
      log.info({ req, res, info: `User ${user.email} deleted` }, `User ${user.email} deleted`);
      return res.status(200).send({ success: `User ${user.email} successfully deleted` });
    }

    return res.status(404).send(`User ${id} not found`);
  } catch (err) {
    log.error({ req, res, err }, `Error deleting user ${id}`);
    return res.status(500).send({ error: err });
  }
}

/**
 * Creates search query for mongoose User model
 *
 * @param {String} query
 * @param {Object} startDate - Date object
 * @param {Object} endDate - Date object
 * @return {Object} searchQuery - Mongoose search query for user
 */
async function prepareMongooseDataBaseQuery(query, startDate, endDate) {
  const searchQuery = {};
  const $or = [];
  const $and = [];

  if (query) {
    // Check if searchQuery is valid mongo ObjectId
    if (query.match(/^[0-9a-fA-F]{24}$/)) {
      // Search query by id
      $or.push({ _id: mongoose.Types.ObjectId(query) });
    } else {
      const reExp = new RegExp(query, 'i');
      // Search query for for other User properties
      $or.push({ name: reExp }, { email: reExp }, { roles: reExp });
    }
  }

  if (startDate && endDate) {
    const gte = await parseDateYearMonthDay(startDate);
    const lte = await parseDateYearMonthDay(endDate);
    // Search query for text search and date range
    $and.push({
      $or: [
        {
          createdAt: { $gte: gte, $lte: lte },
        },
        {
          dateOfBirth: { $gte: gte, $lte: lte },
        },
      ],
    });
  }

  if ($and.length > 0 && $or.length > 0) {
    $and.push({ $or });
    searchQuery.$and = $and;
  } else if ($and.length > 0) {
    searchQuery.$and = $and;
  } else if ($or.length > 0) {
    searchQuery.$or = $or;
  }

  return searchQuery;
}

/**
 * Get user by search query
 *
 * @param {Object} req
 * @param {Object} res
 * @returns {Object} res
 * @author Snær Seljan Þóroddsson
 */
export async function getUser(req, res) {
  const { page, limit, startDate, endDate } = req.query;
  const { query } = req.params;

  if (!query) {
    return res.status(422).send({
      error: 'Parameter query is required to perform user search',
    });
  }
  // Get default pagination object
  const pagination = createPaginationObject(page, limit);
  // Only select these properties
  pagination.select = select;

  try {
    const searchQuery = await prepareMongooseDataBaseQuery(query, startDate, endDate);
    // Fetch user by search query from database
    const result = await User.paginate(searchQuery, pagination);
    return res.status(200).send(result);
  } catch (err) {
    log.error({ req, res, err }, `Error getting user by search query ${query}`);
    return res.status(500).send({ error: err });
  }
}

/**
 * Get all users
 *
 * @param {Object} req
 * @param {Object} res
 * @returns {Object} res
 * @author Snær Seljan Þóroddsson
 */
export async function getUsers(req, res) {
  const { limit, name, email, dateOfBirth, createdAt, page, query, startDate, endDate } = req.query;

  // Get default pagination object
  let pagination = createPaginationObject(page, limit);
  // Enrich pagination object
  pagination = addToPaginationObject(pagination, name, email, dateOfBirth, createdAt);
  // Only select these properties
  pagination.select = select;

  try {
    const searchQuery = await prepareMongooseDataBaseQuery(query, startDate, endDate);
    // Fetch user by search query from database
    // Fetch Users from database
    const result = await User.paginate(searchQuery, pagination);
    return res.status(200).send(result);
  } catch (err) {
    log.error({ req, res, err }, 'Error getting users');
    return res.status(500).send({ error: err.message || err });
  }
}

/**
 * Updates user without comparing password and save to database
 *
 * @param {Object} req
 * @param {Object} res
 * @returns {Object} res
 * @author Snær Seljan Þóroddsson
 */
export async function updateUser(req, res) {
  if (!req.body) return res.status(422).send('No post data found');
  if (!req.user) return res.status(422).send('No user found');

  const { email, password, name, dateOfBirth, phone, roles } = req.body;
  // Validate post request inputs
  const error = validateSignup({
    email,
    password,
    newPassword: null,
    name,
    dateOfBirth,
    phone,
  });

  if (!error) {
    try {
      const $or = [
        { email },
        { 'facebook.email': email },
        { 'twitter.email': email },
        { 'google.email': email },
      ];

      const $set = {};
      if (email) $set.email = email;
      if (password) $set.password = password;
      if (name) $set.name = name;
      if (dateOfBirth) $set.dateOfBirth = dateOfBirth;
      if (phone) $set.phone = phone;
      if (roles) {
        $set.roles = createUserRoles(roles);
      }

      $set.updatedAt = new Date();

      // Find user by email and populate user props
      const updatedUser = await User.findOneAndUpdate({ $or }, { $set }, { new: true });

      // Log in db
      log.info({ req, res, info: updatedUser }, `User ${updatedUser.email} updated in db`);

      if (req.user.email === updatedUser.email) {
        const test = new User(updatedUser); // ToDO finish this
        const user = removeUserProps(test); // eslint-disable-line
        // Send response object with user token and user information
        return res.status(200).json({
          token: tokenForUser(updatedUser),
          ...user,
        });
      }

      return res.status(200).json(updatedUser);
    } catch (err) {
      log.error({ req, res, err }, 'Error updating user');

      return res.status(422).send({ error: err });
    }
  } else {
    return res.status(422).send({ error });
  }
}
