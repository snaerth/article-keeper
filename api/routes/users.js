import { uploadUserImage, updateUser } from '../controllers/users';

/**
 * Default Admin routes
 * @param {Object} app - Express app referece
 * @param {Func} requireAuth - authentication helper function
 * @returns {undefined}
 */
export default function (app, requireAuth) {
  // Upload images
  app.post('/userimage', requireAuth, uploadUserImage);
  // Update user
  app.put('/user', requireAuth, updateUser);
}
