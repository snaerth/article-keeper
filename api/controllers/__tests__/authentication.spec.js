import mongoose from 'mongoose';
import User from '../../models/user';
import config from '../../config';
import { validateSignup, removeUserProps } from '../authentication';
import { saveUser, checkUserByEmail } from '../users';

const { TEST_DB_URL } = config;
const user = new User({
  name: 'John Doe',
  email: 'john@doe.com',
  password: 'Password1',
});

// Validate signup form fields
describe('Validate signup form fields from post request', () => {
  test('Validates signup fields', () => {
    const isValid = validateSignup({
      email: 'snaerth@gmail.com',
      password: 'ThisIsAPassword1',
      newPassword: 'ThisIsAPassword2',
      name: 'John Smith',
      dateOfBirth: new Date(),
    });

    expect(isValid).toBe(null);
  });
});

// Remove props from mongoose user model
describe('Remove props from mongoose user model', () => {
  const userObj = {
    name: 'John Doe',
    email: 'john@doe.com',
    password: 'Password1',
  };
  const user1 = new User(userObj);
  const user2 = new User(userObj);

  test('Remove props from object', () => {
    expect(user2.toObject()).toHaveProperty('password', 'Password1');
    const newUser = removeUserProps(user1);
    expect(newUser).not.toHaveProperty('password');
  });
});

// Db connect
beforeAll(async (done) => {
  await mongoose.connect(TEST_DB_URL);
  done();
});

// Remove user from db and disconnect from db
afterAll(async (done) => {
  await User.collection.remove();
  await mongoose.disconnect(done);
});

// Save user to database
describe('Save mongoose user model to database', async () => {
  // Save user to db
  test('Save user to db', async () => {
    try {
      const data = await saveUser(user);
      expect(data.name).toBe(user.name);
      expect(data.email).toBe(user.email);
      // Remove user from db
      await User.remove(user);
    } catch (error) {
      expect(error.message).toContain('E11000 duplicate');
    }
  });
});

// Check if user exist in db
describe('Check if user exist', async () => {
  // Save user to db
  test('Find user in db', async () => {
    try {
      // Check if user exist
      const checkUser = await checkUserByEmail(user.email);
      expect(checkUser).toBe(undefined);
      // Save user to db
      await saveUser(user);
      // Check if user exist
      await checkUserByEmail(user.email);
      // Remove user from db
      await User.remove(user);
    } catch (error) {
      expect(error).toThrowErrorMatchingSnapshot();
    }
  });
});
