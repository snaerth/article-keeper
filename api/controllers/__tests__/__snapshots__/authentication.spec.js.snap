// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Statrt authentication tests Check if user exist and find user in db 1`] = `"received is not a function"`;
